File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def recc(self, root, mn, mx)->bool:
        if root==None:
            return True
        if root.left==None and root.right==None:
            return (mn<=root.val and root.val<=mx)
        
        if mn<=root.val and root.val<=mx:
            return self.recc(root.left, mn, root.val-1) and self.recc(root.right, root.val+1, mx)
        else:
            return False
    
    def isValidBST(self, root: TreeNode) -> bool:
        
        inf = int(2**32)
        return self.recc(root, -inf, inf)
        

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    bool check(TreeNode* root, long mn, long mx){
        
        if(!root) return true;
        
        int val = root->val;
        if(val>=mx or val<=mn)return 0;
        
        return (check(root->left, mn, val)
               and check(root->right, val, mx)
               );
    }
    
    bool isValidBST(TreeNode* root) {
        
        return check(root, LONG_MIN, LONG_MAX);
    }
};